rfc/rfc-0000-allow-resources-to-report-status.md

```diff
@@ -0,0 +1,335 @@
+# Meta
+[meta]: #meta
+- Name: Allow Resources to Report Status
+- Start Date: 2022-03-22
+- Author(s): @emmjohnson, @martyspiewak
+- Status: Draft <!-- Acceptable values: Draft, Approved, On Hold, Superseded -->
+- RFC Pull Request: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+The RFC proposes modifying the existing templates to allow template authors to indicate how Cartographer can read the state of the resource so that it can be reflected on the owner status.
+
+# Motivation
+[motivation]: #motivation
+
+- Why should we do this?
+  - Currently, a blueprint has no knowledge of the current status of its underlying resources. This RFC proposes how resources can report (and categorize) this information strictly for informational purposes.
+- What use cases does it support?
+  - This feature will enable developers to learn about the progression through the blueprint by looking at the owner status alone. It will also surface information about the health of the underlying resources directly on the owner.
+  - This will also enable integrations that may be looking to create a UI around Cartographer, allowing the UI to display the health of the blueprint without having to make an API call to k8s to retrieve the status of every resource, and without having to know how to define health for every stamped object type.
+- What is the expected outcome?
+  - Users will have more insight into the progression of the blueprint and the health of each underlying resource.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes creating a new CRD, `ClusterResourceStatusRule`. Template authors use this CRD to define the conditions or fields on a stamped resource that determine whether the resource is succeeding or failing. The template author then references this `ClusterResourceStatusRule` from their template.
+
+Cartographer will attempt to read the specified conditions and/or fields and reflect this state on the owner status. If Cartographer cannot determine success or failure, it will show the state as unknown.
+
+```yaml
+---
+apiVersion: carto.run/v1alpha1
+kind: ClusterResourceStatusRule
+metadata:
+  name: ready-rule
+spec:
+  healthy:
+    matchConditions:
+      - type: Ready
+        status: 'True'
+  unhealthy:
+    matchConditions:
+      - type: Ready
+        status: 'False'
```

Comment:
For resources that expose an aggregate condition, it would be nice to reference that condition type once and not need to distinguish between True/False/Unknown. Resource that split the health across multiple conditions  are less common than those that aggregate to a single condition.

Reply:
Adopted this suggestion in most recent updates to the RFC.